File: src/main/java/org/simbrain/world/odorworld/sensors/TileSensor.java
package org.simbrain.world.odorworld.sensors;

import org.simbrain.util.UserParameter;
import org.simbrain.util.piccolo.TileMap;
import org.simbrain.workspace.Producible;
import org.simbrain.world.odorworld.entities.OdorWorldEntity;

public class TileSensor extends Sensor {

    /**
     * Current value of the sensor.
     */
    private double value = 0;

    /**
     * The amount of activation when this sensor is activated
     */
    @UserParameter(label = "Output Amount",
            description = "The amount of activation to be sent to a neuron coupled with this sensor.",
            order = 3)
    private double outputAmount = 1;

    /**
     * The tile id of the tile to sense
     */
    @UserParameter(label = "Tile ID",
            description = "The ID of the tile this sensor should sense.",
            order = 4)
    private int tileIdToSense = 0;

    /**
     * Construct a tile sensor to sense a specific tile of the given ID.
     *
     * @param tileId the tile id of the tile to sense
     */
    public TileSensor(int tileId) {
        this.tileIdToSense = tileId;
    }

    public TileSensor() {
        super("Tile Sensor");
    }

    /**
     * Construct a copy of a tile sensor.
     *
     * @param tileSensor the tile sensor to copy
     */
    public TileSensor(TileSensor tileSensor) {
        super(tileSensor);
        this.outputAmount = tileSensor.outputAmount;
        this.tileIdToSense = tileSensor.tileIdToSense;
    }

    @Override
    public void update(OdorWorldEntity parent) {
        value = 0;

        TileMap tileMap = parent.getWorld().getTileMap();

        if (tileMap.hasTileIdAtPixel(tileIdToSense, parent.getLocation().getX(), parent.getLocation().getY())) {
            value = outputAmount;
        }
    }

    @Producible(customDescriptionMethod = "getAttributeDescription")
    public double getCurrentValue() {
        return value;
    }

    @Override
    public String getName() {
        return "Tile Sensor";
    }

    @Override
    public TileSensor copy() {
        return new TileSensor(this);
    }

    @Override
    public String getLabel() {
        if (super.getLabel().isEmpty()) {
            return "Sense tile id " + tileIdToSense;
        } else {
            return super.getLabel();
        }
    }
}

File: src/main/java/org/simbrain/world/odorworld/sensors/TileSensor.kt
package org.simbrain.world.odorworld.sensors

import org.simbrain.util.UserParameter
import org.simbrain.util.decayfunctions.DecayFunction
import org.simbrain.util.decayfunctions.LinearDecayFunction
import org.simbrain.util.piccolo.*
import org.simbrain.util.plus
import org.simbrain.world.odorworld.entities.OdorWorldEntity

/**
 * Sensor that reacts when an object of a given type is near it.
 * <br></br>
 * While the smell framework involves objects emitting smells, object type
 * sensors have a sensitivity, and are more "sensor" or "subject" based than
 * object based.
 * <br></br>
 * The sensor itself is currently fixed at the center of the agent. We may
 * make the location editable at some point, if use-cases emerge.
 */
class TileSensor @JvmOverloads constructor(
    @UserParameter(label = "Tile Type", description = "What type of tile this sensor responds to", order = 3)
    var tileType: String = "water",
    radius: Double = DEFAULT_RADIUS,
    angle: Double = DEFAULT_THETA
) : SensorWithRelativeLocation(angle, radius), VisualizableEntityAttribute, WithDispersion {

    /**
     * Decay function
     */
    @UserParameter(label = "Decay Function", isObjectType = true, showDetails = false, order = 15)
    override var decayFunction: DecayFunction = LinearDecayFunction(70.0)
        set(value) {
            field = value
            relativeGridCoordinates = null // invalidate pre-computed grid coordinates
        }

    /**
     * Cached relative grid coordinates this sensor should check
     */
    private var relativeGridCoordinates: List<GridCoordinate>? = null

    override var showDispersion = false

    override fun update(parent: OdorWorldEntity) {
        currentValue = 0.0
        val sensorLocation = computeAbsoluteLocation(parent)
        currentValue = with(parent.world.tileMap) {
            (relativeGridCoordinates ?: getRelativeGridLocationsInRadius(decayFunction.dispersion).toList().also { relativeGridCoordinates = it })
                .map { it + sensorLocation.asPixelCoordinate().toGridCoordinate() }
                .map { it.asGridCoordinate() to getTileStackAt(it.x.toInt(), it.y.toInt()) }
                .filter { (_, tiles) -> tiles.any { it.type == tileType } }
                .map { (pos) -> pos.toPixelCoordinate().distance(sensorLocation) }
                .sumOf { decayFunction.getScalingFactor(it) * baseValue }
        }
    }

    override fun copy(): TileSensor {
        return TileSensor().applyCommonCopy().apply {
            tileType = this@TileSensor.tileType
            decayFunction = this@TileSensor.decayFunction.copy() as DecayFunction
        }
    }

    override val name: String
        get() = "Tile Sensor"

    override fun getLabel(): String {
        return if (super.getLabel().isEmpty()) {
            "$directionString$tileType Detector"
        } else {
            super.getLabel()
        }
    }
}